aten/src/ATen/cuda/CUDAApplyUtils.cuh

```diff
@@ -126,41 +133,252 @@ void rearrangeDims(detail::TensorInfo<T1, IndexType>* aInfo,
 #define AT_APPLY_THREADS_PER_BLOCK 32 * 16
 #define AT_APPLY_BLOCKS_PER_SM 4
 
+// The `step` argument is used to support Op that operates on multiple elements
+// at the same time. See NOTE [ CUDA_tensor_applyN helpers ] below for how Op
+// may look like.
+
+template <typename Op,
+          typename scalar,
+          typename IndexType,
+          int ADims,
+          int step,
+          typename... Offsets>
+struct ApplyOp1 {
+__device__ __forceinline__
+static void apply(detail::TensorInfo<scalar, IndexType> &a, const Op &op, int n,
+                  IndexType linearIndex, Offsets... aOffsets) {
+  // Convert `linearIndex` into an offset of `a`
+  const IndexType aOffset = sizeof...(Offsets) < n ?
+    detail::IndexToOffset<scalar, IndexType, ADims>::get(linearIndex, a) : 0;
+
+  ApplyOp1<Op, scalar, IndexType, ADims, step - 1, const IndexType, Offsets...>::apply(
+    a, op, n, linearIndex + 1, aOffsets..., aOffset
+  );
+}
+};
+
+// Specialize step=1 case. We don't need to pass in how many elements need to
```

Comment:
You say `step=1`, but I see `0` in the template parameter...

Reply:
Sorry, it's misnamed. Here the `step` is different from the `step` arg in `CUDA_tensor_applyN`. It's more like `remaining_step`. I will change them. The `step=1` comes from that we have `remaining_step=0` and only have one `typename Offset ` (vs. `typename... Offsets` below).

Reply:
Perhaps you mean, the single offset case?

Reply:
Yep, I'll clarify this.